### PR TITLE
Optimize Syntax.Util destruct_typ_as_formula

### PR DESCRIPTION
**Pull request by ctk21**
This PR optimizes the `destruct_typ_as_formula` function in `FStar.Syntax.Util.fst` which was seen as hot in some profiling runs of ulib files (and a couple from HACL*).

The PR removes an unused list `u_connectives` and refactors the list `connectives`. These lists, while constant by intent, do not extract to constants in OCaml because the integers get parsed by C stubs  into Zarith big-integers. The PR also reworks (and unifies) the lookup algorithms for `destruct_base_conn` and `destruct_sq_base_conn`; primarily to lookup first by arity then search by string (this turned out faster than using a match). There is a tension between optimization, maintainability and clarity. I hope that the balance is ok.

Benchmarking the code with these changes I get:
```
                  Old                               New
                  n       total   user    system    n       total   user    system
micro-benchmarks  61      55.68   51.61   3.839     61      52.61   48.74   3.626
ocaml_extract     38      117.7   113.5   4.146     38      118.2   113.9   4.245
ulib              208     380     359.8   19.5      208     330.6   310.6   19.23
```

FStar.ModifiesGen.fst: 35.2s -> 31.4s (~10%)
There is a cost as ocaml_extract is a tiny bit slower (~0.5%) across many files

Couple of things to note:
 - the (unused) `u_connective` list seems at odds with the match statement below it in the existing code, `PC.and_lid` is mapping to itself in the match but `PC.c_and_lid` in `u_connective`. This PR takes the match behaviour which is what executes.
 - there can be a slight 0.5% runtime degradation on some files; for example the ocaml_extract set. This feels more than offset by the improvement in ulib.

### CODE REVIEW

File: src/syntax/FStar.Syntax.Util.fs [hunk 1]
```diff
@@ -1362,37 +1362,56 @@ type connective =
...
-            let t = un_uinst t in
-            if is_constructor t lid
-            && List.length args = arity
-            then Some (BaseConn(lid, args))
-            else None in
-        U.find_map connectives (aux f) in
+    let destruct_base_conn f =
+        let t, args = head_and_args (unmeta_monadic f) in
+        let t = un_uinst t in
+        match (pre_typ t).n with
```
Comment (msprotz) [inline src/syntax/FStar.Syntax.Util.fs]: I'd like to get a set of expert eyes here on the usage of `pre_typ` and why it's needed. I'm not familiar with this helper function, and don't know whether e.g. `compress` would be more suitable.

@nikswamy @mtzguido

**Comment by msprotz** [comment]
I reviewed this and the changes make sense to me. I don't find the cost in readability to be excessive.

One avenue for improvement is to unify the two tables (at the expense of being more verbose for the base connectors which don't change the `lid`), and make them a single function, which in turn would eliminate a bunch of code and might even be faster.

Asking @nikswamy and @mtzguido for more feedback here. Thanks.

**Comment by ctk21** [comment]
> One avenue for improvement is to unify the two tables (at the expense of being more verbose for the base connectors which don't change the `lid`), and make them a single function, which in turn would eliminate a bunch of code and might even be faster.

Thanks for the suggestion, I've made this change. I think it is better code and benchmarking shows it's a bit faster (~0.5-1%) than before the table lookups were unified.

File: src/syntax/FStar.Syntax.Util.fs (hunk 1 above)
Comment (mtzguido) [reply src/syntax/FStar.Syntax.Util.fs]: Yeah, this looks odd and different to what `destruct_sq_base_conn` is doing, I'm trying out a CI build just using `compress`.

Comment (mtzguido) [reply src/syntax/FStar.Syntax.Util.fs]: Got a green in `guido_test1827`, I would add the changes there to this PR and merge.

**Comment by nikswamy** [comment]
This code is clearly trying to remove several kinds of meta data on ast nodes before analysis.

Unmeta_monadic is removing effect annotations.

Unuinst is removing universe instantiations

Pre_typ is removing ascriptions and refinements.

And there's compress, of course.

Aside from compress, it's not clear to me that any of this is actually needed.

Which, if any, are you retaining, Guido?

**Comment by mtzguido** [comment]
I only left `un_uinst`, since I suspected that one makes sense, but not the others: this was my change https://github.com/FStarLang/FStar/commit/67189ab245b524440eb25c3814d0d47864f21f16. (Will try now without the `un_uinst`, and report back.)

**Comment by mtzguido** [comment]
Yup, some early failures in `ulib/`, e.g. `FStar.Real.fsti`. The diff in `smt2` files shows fragments like this:
```diff
 ;; def=prims.fst(425,14-425,15); use=prims.fst(425,14-425,15)
-(= (Prims.b2t (Prims.op_Equality (FStar.Real.real Dummy_value)
+(Valid 
+;; def=prims.fst(425,14-425,15); use=prims.fst(425,14-425,15)
+(Prims.eq2 Tm_type
+(Prims.b2t (Prims.op_Equality (FStar.Real.real Dummy_value)
 (BoxReal (/ 1.0
 (BoxReal_proj_0 (FStar.Real.sqrt_2 Dummy_value))))
 (BoxReal (/ (BoxReal_proj_0 (FStar.Real.sqrt_2 Dummy_value))
 2.0))))
 @x3)
 )
+)
```
So without the `un_uinst`, `destruct_typ_as_formula` doesn't properly trigger on equalities.

**Comment by ctk21** [comment]
Thanks Guido, I've updated the PR to remove the `pre_typ` but leave the `un_uninst` as in your patch.

**Comment by mtzguido** [comment]
Already merged this manually, let's see if github picks it up...

**Pull request merged**